data_structures/avltree.cpp

```diff
@@ -1,198 +1,426 @@
+    std::cout.rdbuf(old_cout);
+    std::string output = oss.str();
+    output.erase(std::remove_if(output.begin(), output.end(), ::isspace),
+                 output.end());
+    assert(output == "4356");
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing tree balancing.
+ *
+ * @returns `void`
+ */
+static void test_balancing() {
+    std::cout << "Testing AVL balancing...";
+
+    AVLTree tree;
+    tree.insert(1);
+    tree.insert(2);
+    tree.insert(3);
+    tree.insert(4);
+    tree.insert(5);
+
+    std::ostringstream oss;
+    std::streambuf* old_cout = std::cout.rdbuf(oss.rdbuf());
+    tree.levelOrder();
+    std::cout.rdbuf(old_cout);
+    std::string output = oss.str();
+    output.erase(std::remove_if(output.begin(), output.end(), ::isspace),
+                 output.end());
+    assert(output == "21435");
+
+    tree.insert(6);
+    tree.insert(7);
+
+    oss.str("");
+    oss.clear();
+    std::cout.rdbuf(oss.rdbuf());
+    tree.levelOrder();
+    std::cout.rdbuf(old_cout);
+    output = oss.str();
+    output.erase(std::remove_if(output.begin(), output.end(), ::isspace),
+                 output.end());
+    assert(output == "4261357");
+
+    std::cout << "ok" << std::endl;
+}
+
+/**
+ * @brief Function for testing edge cases.
+ *
+ * @returns `void`
+ */
+static void test_edge_cases() {
+    std::cout << "Testing AVL edge cases...";
+
+    AVLTree tree;
+
+    // Deleting from an empty tree
+    tree.deleteNode(1);
+    std::ostringstream oss;
+    std::streambuf* old_cout = std::cout.rdbuf(oss.rdbuf());
+    tree.levelOrder();
+    std::cout.rdbuf(old_cout);
+    std::string output = oss.str();
+    output.erase(std::remove_if(output.begin(), output.end(), ::isspace),
+                 output.end());
+    assert(output == "");
+
+    // Inserting duplicate elements
+    tree.insert(1);
+    tree.insert(1);
+    oss.str("");
+    oss.clear();
+    std::cout.rdbuf(oss.rdbuf());
+    tree.levelOrder();
+    std::cout.rdbuf(old_cout);
+    output = oss.str();
+    output.erase(std::remove_if(output.begin(), output.end(), ::isspace),
+                 output.end());
+    assert(output == "1");
+
+    std::cout << "ok" << std::endl;
 }
 
 /**
- * @brief Main function
+ * @brief Main function for running tests.
+ *
  * @returns 0 on exit
  */
 int main() {
-    // Testing AVL Tree
-    node *root = nullptr;
-    int i = 0;
-    for (i = 1; i <= 7; i++) root = insert(root, i);
-    std::cout << "LevelOrder: ";
-    levelOrder(root);
-    root = deleteNode(root, 1);  // Deleting key with value 1
-    std::cout << "\nLevelOrder: ";
-    levelOrder(root);
-    root = deleteNode(root, 4);  // Deletin key with value 4
-    std::cout << "\nLevelOrder: ";
-    levelOrder(root);
-    deleteAllNodes(root);
+    test_insert();
+    test_deleteNode();
+    test_levelOrder();
+    test_balancing();
+    test_edge_cases();
+    std::cout << "All tests passed!" << std::endl;
     return 0;
-}
+}
```

Comment:
```suggestion
}

```

Reply:
did not understand this issue... Is it resolved?